Framework/CurveFitting/src/Functions/CrystalFieldPeakUtils.cpp

```diff
@@ -179,41 +219,61 @@ size_t buildSpectrumFunction(API::CompositeFunction &spectrum,
 /// @param xVec :: x-values of a tabulated width function.
 /// @param yVec :: y-values of a tabulated width function.
 /// @param fwhmVariation :: A variation in the peak width allowed in a fit.
+/// @param defaultFWHM :: A default value for the FWHM to use if xVec and yVec
+///        are empty.
+/// @param fixAllPeaks :: If true fix all peak parameters
 /// @return :: The new number of fitted peaks.
 size_t updateSpectrumFunction(API::CompositeFunction &spectrum,
+                              const std::string &peakShape,
                               const FunctionValues &centresAndIntensities,
                               size_t nOriginalPeaks, size_t iFirst,
                               const std::vector<double> &xVec,
                               const std::vector<double> &yVec,
-                              double fwhmVariation) {
+                              double fwhmVariation, double defaultFWHM,
+                              bool fixAllPeaks) {
   size_t nGoodPeaks = calculateNPeaks(centresAndIntensities);
   size_t maxNPeaks = calculateMaxNPeaks(nGoodPeaks);
+  size_t nFunctions = spectrum.nFunctions();
   bool mustUpdateWidth = !xVec.empty();
+  const bool fixByDefault = true;
 
   for (size_t i = 0; i < maxNPeaks; ++i) {
-    auto fun = spectrum.getFunction(i + iFirst);
-    auto &peak = dynamic_cast<API::IPeakFunction &>(*fun);
-    if (i < nGoodPeaks) {
-      auto centre = centresAndIntensities.getCalculated(i);
-      peak.setCentre(centre);
-      peak.setIntensity(centresAndIntensities.getCalculated(i + nGoodPeaks));
-      if (mustUpdateWidth) {
-        auto fwhm = peak.fwhm();
-        auto expectedFwhm = calculateWidth(centre, xVec, yVec);
-        if (expectedFwhm <= 0.0) {
-          ignorePeak(peak, fwhm);
-        } else if (fabs(fwhm - expectedFwhm) > fwhmVariation) {
-          peak.setFwhm(expectedFwhm);
-          setWidthConstraint(peak, expectedFwhm, fwhmVariation);
+    if (i < nFunctions) {
+      auto fun = spectrum.getFunction(i + iFirst);
+      auto &peak = dynamic_cast<API::IPeakFunction &>(*fun);
+      if (i < nGoodPeaks) {
+        auto centre = centresAndIntensities.getCalculated(i);
+        peak.setCentre(centre);
+        peak.setIntensity(centresAndIntensities.getCalculated(i + nGoodPeaks));
+        if (mustUpdateWidth) {
+          auto fwhm = peak.fwhm();
+          auto expectedFwhm = calculateWidth(centre, xVec, yVec);
+          if (expectedFwhm <= 0.0) {
+            ignorePeak(peak, fwhm);
+          } else if (fabs(fwhm - expectedFwhm) > fwhmVariation) {
+            peak.setFwhm(expectedFwhm);
+            setWidthConstraint(peak, expectedFwhm, fwhmVariation);
+          }
+        }
+        peak.unfixIntensity();
+        peak.fixIntensity(fixByDefault);
+      } else {
+        peak.setHeight(0.0);
+        if (i > nOriginalPeaks) {
+          peak.fixAll(fixByDefault);
         }
       }
```

Comment:
Again this method is a little long. Perhaps we could break this into a few different methods.